pkg/apis/planetscale/v2/vitessshard_methods.go

```diff
@@ -151,3 +153,47 @@ func (s *VitessShardSpec) CellInCluster(cellName string) bool {
 	_, inZoneMap := s.ZoneMap[cellName]
 	return inZoneMap
 }
+
+// NewVitessShardCondition returns a VitessShardCondition object based on the provided type and initial state.
+func NewVitessShardCondition(ty VitessShardConditionType, initState corev1.ConditionStatus) *VitessShardCondition {
+	now := v1.NewTime(time.Now())
+	return &VitessShardCondition{
+		Type:               ty,
+		Status:             initState,
+		LastTransitionTime: &now,
+		Reason:             "initState",
+		Message:            "The initial state for this VitessShardCondition.",
+	}
+}
+
+// ChangeStatus changes the status if the current status is not the same as the new status, and updates the
+// last transition time.
+func (c *VitessShardCondition) ChangeStatus(newStatus corev1.ConditionStatus, reason, message string) {
+	// We should update reason and message regardless of whether the status type is different.
+	c.Reason = reason
+	c.Message = message
+	if c.Status == newStatus {
+		return
+	}
+
+	c.Status = newStatus
+	now := v1.NewTime(time.Now())
+	c.LastTransitionTime = &now
+}
+
+// Duration returns the duration since LastTransitionTime. It represents how long we've been in the current status for
+// this condition.
+func (c *VitessShardCondition) Duration() time.Duration {
```

Comment:
To me, the name `Duration()` feels incomplete. A Condition doesn't feel like something that itself has a duration, so I'm left wondering, "Duration of what?" It's really the duration since the condition's status changed, so maybe `StatusDuration()`?

We should also do something about the case when LastTransitionTime is nil. It seems awkward to expect callers to check some other field (LastTransitionTime) for nil before calling this method, but I also don't know what this method should return if the time is nil (unknown). Any ideas?

Reply:
That's a good name suggestion - I'll change it. As for LastTransitionTime being nil, I think we should return a Duration of 0.

Reply:
Method name update: https://github.com/planetscale/vitess-operator/pull/31/commits/89eb420dbd304ab7b6a9b440d46fc4432a93127f.

Reply:
Added nil check and zero return https://github.com/planetscale/vitess-operator/pull/31/commits/7090ba557e9ebe152b564df4b9180278be40ee57.